p4-16/psa/PSA.mdk

```diff
@@ -405,27 +417,36 @@ numerical translation between controller port id values and data plane
 port id values, for each of the channels of communication between the
 controller and data plane described above.
 
-Occurrences of values of type `PortId_t` or `PortIdInHeader_t` must
-include a `p4runtime_translation` annotation so the compiler can mark
-occurrences of these values appropriately, so the agent can know which
-values need this translation.  See the example P4 header definition
-below.
-
-```
-[INCLUDE=examples/psa-example-parser-error-handling.p4:PortId_Annotation_Example]
-```
+The file `psa.p4` contains an annotation `p4runtime_translation` on
+the `type` definition of types `PortId_t` and `PortIdInHeader_t`.
+This enables the compiler to mark all uses of values of these types
+that are accessible from the P4Runtime API, so the agent software
+knows to translate them.  The benefit is that you do not need to put
+any special markings on your uses of values of these types throughout
+your P4 program.
+
+The cost of this approach is: if you want to do arithmetic on values
+of these types, you must explicitly cast them to a normal `bit<W>`
+type.  The `psa.p4` include file defines `PortIdUint_t` as a `typedef`
+with exactly the same width in bits as `type` `PortId_t`, so you can
+cast values of type `PortId_t` to type `PortIdUint_t`, and then you
+can perform all P4 arithmetic operations on the value.  The result
+must be explicitly cast back to type `PortId_t` if you wish to assign
+it to a metadata field with that type.  Corresponding types with
+`Uint` in their name are defined for all PSA types.
```

Comment:
Can we also have a note to compiler implementers that mentions what should be the recommended behavior if this is not followed?

Reply:
I do not understand what you mean by "if this is not followed".  The definition of `type` in the P4_16 language specification allows you to do nothing with values of those types except: assign them to other values with an identical type, compare them for == or != with values of an identical type, or cast them to the "underlying" type.  You can also cast values from the underlying type to the type.

Any P4_16 compiler that does anything else isn't a P4_16 compiler (or implements an extension not in the spec).

Reply:
I am not changing anything related to this in the next commit, since it is not yet clear to me what should be changed.

Reply:
I missed that `type` is defined in the language as having a restricted set of operations, and thus any compiler should reject arithmetic ops. I was thinking of it more as a `typedef`. Looks good.